### PR TITLE
Dragleave

### PR DESCRIPTION
**Pull request by christophe-g**
Added containerClass, a css class added when dragitem hover on a container; fires sortenter when a dragitem enters a container; fires dragleave when a dragitem leaves a container.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -211,6 +220,15 @@ sortable('.sortable', {
 });
 ```
 
+### containerClass
```
Comment (lukasoppermann) [inline README.md]: I like the idea. What we need to do now is to improve the naming to make it more obvious when this class is added.
What do you think about `dropTargetContainerClass`? I know it is a mouthful. I am open for better ideas, I just want the class to reflect its function more.

File: README.md [hunk 2]
```diff
@@ -211,6 +220,15 @@ sortable('.sortable', {
...
+Use `containerClass` option to apply a css Class to the container. The class is added when dragged item enters the container and removed when it leaves it (or is dropped).
+
+``` javascript
+sortable('.sortable', {
+  containerClass: 'container-hovered' // Defaults to false
```
Comment (lukasoppermann) [inline README.md]: this could be changed to `is-drop-target`

File: src/html5sortable.ts [hunk 3]
```diff
@@ -364,11 +380,33 @@ export default function sortable (sortableElements, options: configuration|objec
             originalTarget: target
           }
         }))
+
+        _on(sortableContainer, 'dragleave', e => {
+          const outTarget =  e.releatedTarget || e.fromElement;
```
Comment (lukasoppermann) [inline src/html5sortable.ts]: What is `outTarget`? Can we rename this to make it more clear?

**Comment by lukasoppermann** [comment]
Hey @christophe-g, I love this PR, a really great addition to the package.

I left a couple comments. Additionally could you please add some tests to cover the new "uncovered/untested" lines? https://coveralls.io/builds/31215801/source?filename=src%2Fhtml5sortable.ts#L368

This is really important so that we can make sure future additions will not break this part of the package.

Thank you. 👍

**Comment by christophe-g** [comment]
Thanks a lot for the answer @lukasoppermann 

Agreed on the comments - will hopefully have some time next week to add tests (and integrate your remarks).
Cheers, 
C.

**Comment by lukasoppermann** [comment]
Great, thank you very much. Looking forward to merge it afterwards. 👍

**Comment by lukasoppermann** [comment]
Hey @christophe-g did you have time to work on this. I would love to merge it once it is updated. 👍

**Comment by lukasoppermann** [comment]
Hey @kaffarell do you think you can get this one patched up so that we can merge it? The way I understand it, it could be a nice addition to the package.

**Comment by kaffarell** [comment]
👍 Ok, will try later today...

**Comment by kaffarell** [comment]
@lukasoppermann I resolved the first two of your comments but I really don't get what @christophe-g meant with `outTarget` I also couldn't write any tests.
Will try tomorrow another time understanding this

File: src/html5sortable.ts [hunk 4]
```diff
@@ -61,12 +61,12 @@ const _removeItemEvents = function (items) {
   _off(items, 'mouseleave')
 }
 
-// remove container eventes 
-const _removeContainerEvents = () => {
-  if(originContainer) {
+// Remove container events
+const _removeContainerEvents = function () {
```
Comment (lukasoppermann) [inline src/html5sortable.ts:65]: Why did you change this to use the `function` keyword?

File: docs/html5sortable.js [hunk 5]
```diff
@@ -1010,6 +1025,25 @@ var sortable = (function () {
                           originalTarget: target
                       }
                   }));
+                  addEventListener(sortableContainer, 'dragleave', function (e) {
+                      var outTarget = e.releatedTarget || e.fromElement;
```
Comment (lukasoppermann) [inline docs/html5sortable.js:1029]: This definitely has to be fixed to `relatedTarget` (there is a typo). `e.fromElement` is for very old browsers, but maybe we just add a comment `// e.fromElement for very old browsers, similar to relatedTarget`.

Maybe you can also add a comment above `// TODO: rename outTarget to be more self-explanatory`-

Basically I think it is the parent element (not sortable) of the sortable that the item was in. We are checking that it actually is moved to a different sortable and does not stay within.

File: src/html5sortable.ts [hunk 6]
```diff
@@ -364,8 +380,27 @@ export default function sortable (sortableElements, options: configuration|objec
             originalTarget: target
           }
         }))
-      }
 
+        _on(sortableContainer, 'dragleave', e => {
+          const outTarget = e.releatedTarget || e.fromElement
```
Comment (lukasoppermann) [inline src/html5sortable.ts]: This definitely has to be fixed to relatedTarget (there is a typo). e.fromElement is for very old browsers, but maybe we just add a comment // e.fromElement for very old browsers, similar to relatedTarget.

Maybe you can also add a comment above // TODO: rename outTarget to be more self-explanatory-

Basically I think it is the parent element (not sortable) of the sortable that the item was in. We are checking that it actually is moved to a different sortable and does not stay within.

File: src/html5sortable.ts (hunk 4 above)
Comment (kaffarell) [reply src/html5sortable.ts:65]: I changed it to be the same as the function above, maybe we could change all functions to arrow functions...

**Comment by lukasoppermann** [comment]
Hey @kaffarell did you also already write some tests for this new update? 
I think if we merge this pr and solve the #548 issue we can create a new release.

**Review by lukasoppermann — APPROVED** [review]

**Comment by lukasoppermann** [comment]
Hey @kaffarell,
I would love to merge this and release a new version. Did you start on the tests? If so do you have a state of them somewhere?

**Comment by kaffarell** [comment]
Sorry for the late reply, had a lot to do in the recent weeks. I fixed the typo and added the comments But unfortunately I was not able to write the tests.

**Comment by lukasoppermann** [comment]
Okay, no worries. Did you manually test it though?

**Comment by kaffarell** [comment]
Tested it manually now and it is working...
